File: packages/silex-server/src/services/jobs.rs
/*
 * Silex website builder, free/libre no-code tool for makers.
 * Copyright (c) 2023 lexoyo and Silex Labs foundation
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or any later version.
 */

//! Job management service
//!
//! Tracks async jobs like publication operations.
//! Jobs can be queried by ID to check their status.

use std::collections::HashMap;
use std::sync::{Arc, RwLock};
use uuid::Uuid;

use crate::models::{JobId, JobStatus, PublicationJobData};

/// Job manager for tracking async operations
///
/// The job manager maintains a registry of active and completed jobs.
/// Jobs are automatically cleaned up after a timeout (configurable).
#[derive(Clone)]
pub struct JobManager {
    /// Map of job ID to job data
    /// Using RwLock for thread-safe access
    jobs: Arc<RwLock<HashMap<JobId, PublicationJobData>>>,
}

impl JobManager {
    /// Create a new job manager
    pub fn new() -> Self {
        JobManager {
            jobs: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// Start a new job
    ///
    /// Creates a new job with a unique ID and IN_PROGRESS status.
    /// Returns the job data with the generated ID.
    pub fn start_job(&self, message: String) -> PublicationJobData {
        let job_id = Uuid::new_v4().to_string();
        let job = PublicationJobData::new(job_id.clone(), message);

        // Store the job in the registry
        let mut jobs = self.jobs.write().unwrap();
        jobs.insert(job_id, job.clone());

        job
    }

    /// Get a job by ID
    ///
    /// Returns None if the job doesn't exist.
    pub fn get_job(&self, job_id: &JobId) -> Option<PublicationJobData> {
        let jobs = self.jobs.read().unwrap();
        jobs.get(job_id).cloned()
    }

    /// Mark a job as completed
    pub fn complete_job(&self, job_id: &JobId) {
        let mut jobs = self.jobs.write().unwrap();
        if let Some(job) = jobs.get_mut(job_id) {
            job.base.status = JobStatus::Success;
            job.end_time = Some(chrono::Utc::now().timestamp_millis());
        }
    }

    /// Mark a job as failed
    pub fn fail_job(&self, job_id: &JobId, error: &str) {
        let mut jobs = self.jobs.write().unwrap();
        if let Some(job) = jobs.get_mut(job_id) {
            job.base.status = JobStatus::Error;
            job.base.message = error.to_string();
            job.error(error.to_string());
            job.end_time = Some(chrono::Utc::now().timestamp_millis());
        }
    }

}

impl Default for JobManager {
    fn default() -> Self {
        Self::new()
    }
}

File: packages/silex-lib/src/ts/server/jobs.ts
/*
 * Silex website builder, free/libre no-code tool for makers.
 * Copyright (c) 2023 lexoyo and Silex Labs foundation
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { JobData, JobStatus, JobId } from '../types'

// Delay after which the job is deleted from memory
const TIME_TO_KEEP_AFTER_STOP = 60*10*1000 // 10 min

// Store the jobs in memory
const jobs = new Array<JobData>()
const timers = new WeakMap<JobData, NodeJS.Timeout>()

// Key which will change on server restart
const processKey = Math.ceil(Math.random() * 1000000)

// Next id for the next job
let nextJobId = 0

// Create a new job
export function startJob(message = ''): JobData {
  const jobId = `${processKey}_${++nextJobId}`
  const job: JobData = {
    jobId,
    status: JobStatus.IN_PROGRESS,
    message: message
  }
  jobs.push(job)
  return job
}

// Retriev a job if it exists
export function getJob(id: JobId): JobData | undefined {
  return jobs.find(job => job.jobId === id)
}


// End the job with success
export function jobSuccess(id: JobId, message = ''): boolean {
  return end(id, JobStatus.SUCCESS, message)
}

// End the job with error
export function jobError(id: JobId, message = ''): boolean {
  return end(id, JobStatus.ERROR, message)
}

// Remove a job from memory
// This is also used in unit tests to avoid memory lea
export function killJob(job: JobData) {
  clearTimeout(timers.get(job))
  timers.delete(job)
  jobs.splice(jobs.findIndex(j => job.jobId === j.jobId), 1)
}

// End the job
function end(id: JobId, status: JobStatus, message = ''): boolean {
  const job = getJob(id)
  if(job) {
    job.status = status
    if(message) job.message = message
    timers.set(job, setTimeout(() => killJob(job), TIME_TO_KEEP_AFTER_STOP))
    return true
  }
  return false
}

// Type definition of the job manager
export interface JobManager {
  startJob: typeof startJob
  jobSuccess: typeof jobSuccess
  jobError: typeof jobError
}

// Export the job manager
export const jobManager: JobManager = { startJob, jobSuccess, jobError }
